modules/core/src/main/java/org/apache/ignite/internal/processors/cache/consistentcut/ConsistentCut.java

```diff
@@ -0,0 +1,260 @@
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.internal.pagemem.wal.record.ConsistentCutFinishRecord;
+import org.apache.ignite.internal.pagemem.wal.record.ConsistentCutStartRecord;
+import org.apache.ignite.internal.pagemem.wal.record.RolloverType;
+import org.apache.ignite.internal.pagemem.wal.record.WALRecord;
+import org.apache.ignite.internal.processors.cache.GridCacheSharedContext;
+import org.apache.ignite.internal.processors.cache.persistence.wal.WALPointer;
+import org.apache.ignite.internal.processors.cache.transactions.IgniteInternalTx;
+import org.apache.ignite.internal.processors.cache.transactions.IgniteTxManager;
+import org.apache.ignite.internal.processors.cache.version.GridCacheVersion;
+import org.apache.ignite.internal.util.future.GridCompoundFuture;
+import org.apache.ignite.internal.util.future.GridFutureAdapter;
+import org.apache.ignite.internal.util.tostring.GridToStringInclude;
+import org.apache.ignite.internal.util.typedef.internal.CU;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteUuid;
+import org.jetbrains.annotations.Nullable;
+
+import static org.apache.ignite.internal.processors.cache.transactions.IgniteInternalTx.FinalizationStatus.RECOVERY_FINISH;
+import static org.apache.ignite.transactions.TransactionState.ACTIVE;
+import static org.apache.ignite.transactions.TransactionState.COMMITTED;
+import static org.apache.ignite.transactions.TransactionState.MARKED_ROLLBACK;
+import static org.apache.ignite.transactions.TransactionState.ROLLED_BACK;
+import static org.apache.ignite.transactions.TransactionState.UNKNOWN;
+
+/**
+ * Describes current Consistent Cut.
+ */
+public class ConsistentCut extends GridFutureAdapter<WALPointer> {
+    /** */
+    private final GridCacheSharedContext<?, ?> cctx;
+
+    /** */
+    private final IgniteLogger log;
+
+    /** ID of Consistent Cut. */
+    private final UUID id;
+
+    /** Set of checked transactions belonging to the BEFORE side. */
+    @GridToStringInclude
+    private Set<GridCacheVersion> beforeCut;
+
+    /** Set of checked transactions belonging to the AFTER side. */
+    @GridToStringInclude
+    private Set<GridCacheVersion> afterCut;
+
+    /** Collection of transactions removed from {@link IgniteTxManager#activeTransactions()}. */
+    private volatile Set<IgniteInternalFuture<IgniteInternalTx>> removedActive = ConcurrentHashMap.newKeySet();
+
+    /** */
+    ConsistentCut(GridCacheSharedContext<?, ?> cctx, UUID id) {
+        this.cctx = cctx;
+        this.id = id;
+
+        log = cctx.logger(ConsistentCut.class);
+    }
+
+    /** */
+    public UUID id() {
+        return id;
+    }
+
+    /**
+     * Inits local Consistent Cut: prepares list of active transactions to check which side of Consistent Cut they belong to.
+     */
+    protected void init() throws IgniteCheckedException {
+        walLog(new ConsistentCutStartRecord(id), false);
```

Comment:
Let's just invoke wal.log here, without unnecessary `walLog` method.

Reply:
The `walLog` method is used in 2 places and it helps avoid code duplication for debug logging, checking WAL manager for null.